Import all the libraries you require in the cell below.


In [ ]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import MinMaxScaler

# Enter your code here
from sklearn.model_selection import train_test_split
from sklearn.dummy import DummyClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    classification_report,
    recall_score,
    precision_score
)
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, VotingClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression

## Implement the Ensemble methods learnt in class and compare their accuarcies.

The dataset you are going to be using for homework is the **Wisconsin Breast Cancer dataset (cancer.csv)**

The dataset contains a total number of 10 features labeled in either benign or malignant classes. The features have 699 instances out of which 16 feature values are missing. The dataset only contains numeric values.

Attribute Information:

1. Sample code number: id number
2. Clump Thickness: 1 - 10
3. Uniformity of Cell Size: 1 - 10
4. Uniformity of Cell Shape: 1 - 10
5. Marginal Adhesion: 1 - 10
6. Single Epithelial Cell Size: 1 - 10
7. Bare Nuclei: 1 - 10
8. Bland Chromatin: 1 - 10
9. Normal Nucleoli: 1 - 10
10. Mitoses: 1 - 10
11. Class: (2 for benign, 4 for malignant) (**target variable**)

For more information: https://archive.ics.uci.edu/ml/datasets/breast+cancer+wisconsin+(original)

### 1. Read the dataset into variable called '**data**' (1 mark)

In [ ]:
pd.set_option('display.max_columns', 100)
# Enter your code here
data = pd.read_csv("cancer.csv")
data

Sample Code Number  Clump Thickness  Uniformity of Cell Size   \
0               1000025                5                         1   
1               1002945                5                         4   
2               1015425                3                         1   
3               1016277                6                         8   
4               1017023                4                         1   
..                  ...              ...                       ...   
694              776715                3                         1   
695              841769                2                         1   
696              888820                5                        10   
697              897471                4                         8   
698              897471                4                         8   

     Uniformity of Cell Shape  Marginal Adhesion  Single Epithelial Cell Size  \
0                           1                  1                            2   
1                           4                  5                            7   
2                           1                  1                            2   
3                           8                  1                            3   
4                           1                  3                            2   
..                        ...                ...                          ...   
694                         1                  1                            3   
695                         1                  1                            2   
696                        10                  3                            7   
697                         6                  4                            3   
698                         8                  5                            4   

    Bland Chromatin  Bare Nuclei  Normal Nucleoli  Mitosis  Class  
0                 1            3                1        1      2  
1                10            3                2        1      2  
2                 2            3                1        1      2  
3                 4            3                7        1      2  
4                 1            3                1        1      2  
..              ...          ...              ...      ...    ...  
694               2            1                1        1      2  
695               1            1                1        1      2  
696               3            8               10        2      4  
697               4           10                6        1      4  
698               5           10                4        1      4  

[699 rows x 11 columns]

### **Preprocessing**: Data needs to be preprocessed before implementing ensemble methods. It is done for you here. 
### Run the below code first and then answer the questions from 2 - 7.

#### Deleting unnecessary columns: The column "Sample code number" is just an indicator and it's of no use in the modeling. So, let's drop it:


In [ ]:
data.drop(['Sample Code Number'],axis = 1, inplace = True)


#### Handling missing values : 
As mentioned earlier, the dataset contains missing values. The column named "Bland Chromatin" contains them. The missing values are represneted as "?". 

Replace those "?"s with 0's and impute them with Mean Imputation

In [ ]:
data['Bland Chromatin']

0       1
1      10
2       2
3       4
4       1
       ..
694     2
695     1
696     3
697     4
698     5
Name: Bland Chromatin, Length: 699, dtype: object

In [ ]:
data['Bland Chromatin'].unique()

array(['1', '10', '2', '4', '3', '9', '7', '?', '5', '8', '6'],
      dtype=object)

In [ ]:
data.replace('?',0, inplace=True)

In [ ]:
data['Bland Chromatin'].unique()

array(['1', '10', '2', '4', '3', '9', '7', 0, '5', '8', '6'], dtype=object)

In [ ]:
# Convert the DataFrame object into NumPy array otherwise you will not be able to impute
# values = data.values
# Now impute it
# imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
# imputedData = imputer.fit_transform(values)

In [ ]:
# Convert the DataFrame object into NumPy array otherwise you will not be able to impute
values = data.values
# Now impute it
imputer = SimpleImputer(missing_values=0, strategy='mean')
imputedData = imputer.fit_transform(values)

#### Normalizing the data:
Ranges of the features of the dataset are not the same. This may cause a problem. A small change in a feature might not affect the other. To address this problem, normalize the ranges of the features to a uniform range, in this case, 0 - 1.

In [ ]:
scaler = MinMaxScaler(feature_range=(0, 1))
normalizedData = scaler.fit_transform(imputedData)
cols = ['Clump Thickness', 'Uniformity of Cell Size', 'Uniformity of Cell Shape', 'Marginal Adhesion', 'Single Epithelial Cell Size', 'Bland Chromatin', 'Bare Nuclei', 'Normal Nucleoli', 'Mitosis','Class']
normalizedData = pd.DataFrame(normalizedData, columns=cols)
print(normalizedData.head())

   Clump Thickness  Uniformity of Cell Size  Uniformity of Cell Shape  \
0         0.444444                 0.000000                  0.000000   
1         0.444444                 0.333333                  0.333333   
2         0.222222                 0.000000                  0.000000   
3         0.555556                 0.777778                  0.777778   
4         0.333333                 0.000000                  0.000000   

   Marginal Adhesion  Single Epithelial Cell Size  Bland Chromatin  \
0           0.000000                     0.111111         0.000000   
1           0.444444                     0.666667         1.000000   
2           0.000000                     0.111111         0.111111   
3           0.000000                     0.222222         0.333333   
4           0.222222                     0.111111         0.000000   

   Bare Nuclei  Normal Nucleoli  Mitosis  Class  
0     0.222222         0.000000      0.0    0.0  
1     0.222222         0.111111      0.0

### Data preprocessing is done and now you will answer the below questions using the **normalizedData**: 

### 2. Split the data into test and training data with test size - 30%. Compute the baseline classification accuracy for X_train. (3 marks)

In [ ]:
y = normalizedData["Class"]
X = normalizedData.copy()
X.drop("Class", axis = 1, inplace = True)
X

Clump Thickness  Uniformity of Cell Size  Uniformity of Cell Shape  \
0           0.444444                 0.000000                  0.000000   
1           0.444444                 0.333333                  0.333333   
2           0.222222                 0.000000                  0.000000   
3           0.555556                 0.777778                  0.777778   
4           0.333333                 0.000000                  0.000000   
..               ...                      ...                       ...   
694         0.222222                 0.000000                  0.000000   
695         0.111111                 0.000000                  0.000000   
696         0.444444                 1.000000                  1.000000   
697         0.333333                 0.777778                  0.555556   
698         0.333333                 0.777778                  0.777778   

     Marginal Adhesion  Single Epithelial Cell Size  Bland Chromatin  \
0             0.000000                     0.111111         0.000000   
1             0.444444                     0.666667         1.000000   
2             0.000000                     0.111111         0.111111   
3             0.000000                     0.222222         0.333333   
4             0.222222                     0.111111         0.000000   
..                 ...                          ...              ...   
694           0.000000                     0.222222         0.111111   
695           0.000000                     0.111111         0.000000   
696           0.222222                     0.666667         0.222222   
697           0.333333                     0.222222         0.333333   
698           0.444444                     0.333333         0.444444   

     Bare Nuclei  Normal Nucleoli   Mitosis  
0       0.222222         0.000000  0.000000  
1       0.222222         0.111111  0.000000  
2       0.222222         0.000000  0.000000  
3       0.222222         0.666667  0.000000  
4       0.222222         0.000000  0.000000  
..           ...              ...       ...  
694     0.000000         0.000000  0.000000  
695     0.000000         0.000000  0.000000  
696     0.777778         1.000000  0.111111  
697     1.000000         0.555556  0.000000  
698     1.000000         0.333333  0.000000  

[699 rows x 9 columns]

In [ ]:
# Enter your code here
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [ ]:
dummy_classifier = DummyClassifier(strategy='most_frequent')
dummy_classifier.fit(X_train,y_train)
baseline_acc = dummy_classifier.score(X_train,y_train)


### For verifying answer:
print("Baseline Accuracy = ", baseline_acc)

Baseline Accuracy =  0.6441717791411042


### 3.  Bagging : Build a generic Bagging ensemble and print the accuracy (4 marks)
---


Hyperparameters:

Base estimator = DecisionTreeClassifier

n_estimators = 10

random_state = 42

---


In [ ]:
# Generic Bagging model
# Enter your code here
# According to scikitlearn the default base estimator is decision tree classifier
model_bagging = BaggingClassifier(n_estimators = 10, random_state = 42)
model_bagging.fit(X_train, y_train)
pred_bagging = model_bagging.predict(X_test)
acc_bagging = accuracy_score(y_test, pred_bagging)

print(' Accuracy = ', acc_bagging)

 Accuracy =  0.9428571428571428


### 4. RandomForest : (7 marks)
#### a) Build a Random Forest model and print the accuracy (4 marks)
---

Constructor arguments: 


n_estimators = 100, max_features = 7 and random_state = 42 


---




In [ ]:
# Random Forest model
# Enter your code here
model_rf = RandomForestClassifier(n_estimators=100, max_features=7, random_state=42)
model_rf.fit(X_train, y_train)
predict_rf = model_rf.predict(X_test)
recall_rf = recall_score(y_test, predict_rf)
precision_rf = precision_score(y_test, predict_rf)
print(' Recall = ', recall_rf, 'Precision = ', precision_rf)

 Recall =  0.9253731343283582 Precision =  0.9393939393939394


In [ ]:
print("accuracy = ", accuracy_score(y_test, predict_rf))


accuracy =  0.9571428571428572


####  b) Calculate the top 3 important features for the above **RandomForest** model and print them (3 marks)

In [ ]:
# Top 3 features for RandomForest
# Enter your code here
feature_importances = model_rf.feature_importances_
features = X_train.columns
df = pd.DataFrame({'features': features, 'importance': feature_importances}).nlargest(3, 'importance')
print(df)


                   features  importance
1   Uniformity of Cell Size    0.414199
2  Uniformity of Cell Shape    0.280962
5           Bland Chromatin    0.115247


### 5. Boosting: (7 marks)
#### a) Build an AdaBoost model with training data and print the accuracy (4 marks)
---

Hyperparameters:

Base estimator = DecisionTreeClassifier, max_depth = 4

n_estimators = 200

random_state = 42

learning_rate = 0.05


---









In [ ]:
# AdaBoost Classification
# Enter your code here
base_est = DecisionTreeClassifier (max_depth =4)
ada_boost1 = AdaBoostClassifier(base_est, n_estimators=200, random_state=42, learning_rate=.05)
ada_boost1.fit(X_train, y_train)
res1 = accuracy_score(y_test, ada_boost1.predict(X_test))
print('Accuracy:\t {}'.format(res1))

Accuracy:	 0.9666666666666667


#### b) Calculate the top 3 important features for the above **AdaBoost** model and print them (3 marks)

In [ ]:
# Top 3 features for AdaBoost
# Enter your code here
feature_importances = ada_boost1.feature_importances_
features = X_train.columns
df = pd.DataFrame({'features': features, 'importance': feature_importances}).nlargest(3, 'importance')
print(df)


                      features  importance
7              Normal Nucleoli    0.293075
0              Clump Thickness    0.186162
4  Single Epithelial Cell Size    0.111004


### 6. Voting : Using a voting classifier, build an ensemble of RandomForestClassifier, DecisionTreeClassifier, Support Vector Machine and Logistic Regression. (7 marks)


---


Use max_depth = 4, n_estimators = 200, voting = soft

In [ ]:
# Voting Ensemble for Classification
# Enter your code here
clf = DecisionTreeClassifier(random_state=0, max_depth = 4)
rfClf = RandomForestClassifier(n_estimators=200, random_state=0, max_depth=4) 
svmClf = SVC(probability=True, random_state=0) # probability calculation
logClf = LogisticRegression(random_state=0)

# constructing the ensemble classifier by mentioning the individual classifiers.
clf2 = VotingClassifier(estimators = [('rf',rfClf), ('svm',svmClf), ('log', logClf), ('dec', clf)], voting='soft') 

# train the ensemble classifier
clf2.fit(X_train, y_train)

clf2_pred = clf2.predict(X_test)
recall_voting = recall_score(y_test, clf2_pred)
precision_voting = precision_score(y_test, clf2_pred)
print('Accuracy score', accuracy_score(y_test, clf2_pred))

Accuracy score 0.9666666666666667


### 7. Mention the best model among the above 4 models and its accuracy (1 mark)

There is a tie between boosting and voting for being the most accurate (.9666666667)